server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,170 @@
     }
 
-    private void awaitBootstrap(final BootstrapConfiguration bootstrapConfiguration) {
-        if (running == false) {
-            logger.debug("awaitBootstrap: not running");
-            return;
-        }
+    private boolean checkWaitRequirements(Set<DiscoveryNode> nodes) {
+        final Set<DiscoveryNode> selectedNodes = new HashSet<>();
+        for (final String requirement : initialMasterNodes) {
+            final Set<DiscoveryNode> matchingNodes
+                = nodes.stream().filter(n -> matchesRequirement(n, requirement)).collect(Collectors.toSet());
 
-        BootstrapClusterRequest request = new BootstrapClusterRequest(bootstrapConfiguration);
-        logger.trace("sending {}", request);
-        transportService.sendRequest(transportService.getLocalNode(), BootstrapClusterAction.NAME, request,
-            new TransportResponseHandler<BootstrapClusterResponse>() {
-                @Override
-                public void handleResponse(BootstrapClusterResponse response) {
-                    logger.debug("automatic cluster bootstrapping successful: received {}", response);
-                }
-
-                @Override
-                public void handleException(TransportException exp) {
-                    // log a warning since a failure here indicates a bad problem, such as:
-                    // - bootstrap configuration resolution failed (e.g. discovered nodes no longer match those in the bootstrap config)
-                    // - discovered nodes no longer form a quorum in the bootstrap config
-                    logger.warn(new ParameterizedMessage("automatic cluster bootstrapping failed, retrying [{}]",
-                        bootstrapConfiguration.getNodeDescriptions()), exp);
-
-                    // There's not really much else we can do apart from retry and hope that the problem goes away. The retry is delayed
-                    // since a tight loop here is unlikely to help.
-                    transportService.getThreadPool().scheduleUnlessShuttingDown(TimeValue.timeValueSeconds(10), Names.SAME, new Runnable() {
-                        @Override
-                        public void run() {
-                            // TODO: remove the following line once schedule method properly preserves thread context
-                            transportService.getThreadPool().getThreadContext().markAsSystemContext();
-                            awaitBootstrap(bootstrapConfiguration);
-                        }
-
-                        @Override
-                        public String toString() {
-                            return "retry bootstrapping with " + bootstrapConfiguration.getNodeDescriptions();
-                        }
-                    });
-                }
+            if (matchingNodes.isEmpty()) {
+                return false;
+            }
+            if (matchingNodes.size() > 1) {
+                bootstrappingPermitted.set(false);
```

Comment:
I find it confusing to understand where this flag is being set / unset, in particular because this method initially looks to be just a pure predicate, but also has this nasty side-effect here. Can you move this to the caller of this method perhaps?